File: 2017/kotlin/day17.kt
import java.io.File

fun main() {
    val input = File("../__in/17.txt").readText().toInt()

    val buffer = IntArray(2018)
    var pos = 0
    
    for (n in 1..2017) {
        repeat(input) { pos = buffer[pos] }
        buffer[n] = buffer[pos]
        buffer[pos] = n
        pos = buffer[pos]
    }
    println("Part 1: ${buffer[2017]}")

    var target = 0
    pos = 0
    
    for (n in 1..50000000) {
        pos = (pos + input) % n + 1
        if (pos == 1) target = n
    }
    println("Part 2: $target")
}

File: 2017/lua/day17.lua
input = tonumber(io.open("17.txt", "r"):read())

buffer, pos = {[0] = 0}, 0

for n = 1, 2017 do
    for i = 1, input do pos = buffer[pos] end
    
    buffer[n], buffer[pos] = buffer[pos], n; pos = buffer[pos] end

current, pos = 0, 0 -- looking up for "index" of 2017 in resulting table to...
while current ~= 2017 do current, pos = buffer[current], pos + 1 end

for n = 2018, 50000000 do -- ...continue from where part 1 stopped for no reason :)
    pos = (pos + input) % n + 1
    if pos == 1 then target = n end end

print(string.format("Part 1: %d", buffer[2017]))
print(string.format("Part 2: %d", target))